### PR TITLE
[DSCP-124] Provide logging context into resource allocation

### PR DESCRIPTION
**Pull request by Martoon-00**
Split resource allocation into two phases, first prepares logging so that in second phase we can use it for other resources allocation and reporting possible errors.

### CODE REVIEW

File: src/Dscp/Educator/Secret/Real/Functions.hs [hunk 1]
```diff
@@ -51,11 +53,13 @@ genStore = do
     return EducatorSecret{..}
 
 -- | Read store under given path.
-readStore :: (MonadIO m, MonadCatch m) => FilePath -> PassPhrase -> m EducatorSecret
+readStore
+    :: (MonadIO m, MonadCatch m, MonadLogging m)
```
Comment (flyingleafe) [inline src/Dscp/Educator/Secret/Real/Functions.hs]: Why did you add `MonadLogging` here, if you didn't add any actual logging? :thinking:

File: src/Dscp/Resource/Functions.hs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    eres <- try $ runComponentM initDesc (allocInitResources params) $
+      \initCtx -> do
+         eres <- try $ runComponentR desc initCtx component main
+         whenLeft eres $ \(err :: ComponentError) -> do
+             runRIO initCtx $ logError (""+||Doc.pretty err||+"")
+         either throwM pure eres
+    whenLeft eres $ \case
+        ComponentRuntimeFailed{ componentErrorOriginalException = cause }
+            | isJust (fromException @ComponentError cause) ->
+            -- already printed error above
```
Comment (flyingleafe) [inline src/Dscp/Resource/Functions.hs]: Is it possible to have something _different_ but `ComponentError` inside `ComponentRuntimeFailed`? If it is, then we'll silently pass an error in this case.

File: src/Dscp/Resource/Functions.hs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+import Dscp.Launcher.Rio (runRIO)
+import Dscp.Resource.Class (InitContext (..), InitParams (..))
+import Dscp.Resource.Logging (allocLogging)
+
+-- | Catches and prints possible synchronous exceptions thrown
+-- in 'runComponentM'. This includes both resource allocation errors and
+-- application errors.
+handleComponentErrors :: IO a -> IO (Either ComponentError a)
+handleComponentErrors action = do
```
Comment (flyingleafe) [inline src/Dscp/Resource/Functions.hs]: This function seems to be kinda useless, as `runResourceAllocation` does it manually (for reasons).

File: src/Dscp/Educator/Secret/Real/Functions.hs (hunk 1 above)
Comment (Martoon-00) [reply src/Dscp/Educator/Secret/Real/Functions.hs]: `ensureModeIs` can report upon fixing file mode

File: src/Dscp/Resource/Functions.hs (hunk 3 above)
Comment (Martoon-00) [reply src/Dscp/Resource/Functions.hs]: Right :man_facepalming:

File: src/Dscp/Educator/Secret/Real/Functions.hs (hunk 1 above)
Comment (flyingleafe) [reply src/Dscp/Educator/Secret/Real/Functions.hs]: Ok, sorry, missed it.

File: src/Dscp/Resource/Functions.hs (hunk 2 above)
Comment (Martoon-00) [reply src/Dscp/Resource/Functions.hs]: Won't control pass to the second `case` branch is situation you describe?

Comment (flyingleafe) [reply src/Dscp/Resource/Functions.hs]: @Martoon-00 yep sorry, that was stupid :man_facepalming:
